File: BEFoundation Source/NSNumber+BExtension.m
/*!
 @file			NSNumber+BExtension.m
 @copyright		-© 2025 Delicense - @belisoful. All rights released.
 @date			2025-01-01
 @abstract
 @discussion
*/

#import "NSNumber+BExtension.h"


// Efficient power function for Int64 using exponentiation by squaring
// Returns base^exponent, or 0 on overflow
// Note: Negative exponents are not supported (would require floating point)
int64_t pow_int64(int64_t base, uint64_t exponent)
{
	// Handle special cases
	if (exponent == 0) return 1;
	if (base == 0) return 0;
	if (base == 1) return 1;
	if (base == -1) return (exponent & 1) ? -1 : 1;  // (-1)^even = 1, (-1)^odd = -1
	if (exponent == 1) return base;
	
	// Handle negative base
	bool negative_result = (base < 0) && (exponent & 1);
	
	// Work with absolute value to avoid overflow issues during computation
	uint64_t abs_base = (base == INT64_MIN) ? (uint64_t)INT64_MAX + 1 : (base < 0) ? -base : base;
	uint64_t result = 1;
	uint64_t current_base = abs_base;
	
	// Determine overflow limits based on final sign
	uint64_t max_positive = INT64_MAX;
	uint64_t max_negative = (uint64_t)INT64_MAX + 1;  // |INT64_MIN|
	uint64_t overflow_limit = negative_result ? max_negative : max_positive;
	
	while (exponent > 0) {
		uint64_t max_factor = overflow_limit / current_base;
		
		if (exponent & 1) {
			if (result > max_factor) {
				return 0; // Overflow detected
			}
			result *= current_base;
		}
		
		exponent >>= 1;
		
		if (exponent > 0) {
			if (current_base > max_factor) {
				return 0; // Overflow detected
			}
			current_base *= current_base;
		}
	}
	
	// Final sign application
	return negative_result ? -(int64_t)result : (int64_t)result;
}

uint64_t pow_uint64(uint64_t base, uint64_t exponent)
{
	if (exponent == 0) return 1;
	if (base == 0) return 0;
	if (base == 1) return 1;
	if (exponent == 1) return base;

	uint64_t result = 1;
	uint64_t current_base = base;

	while (exponent > 0) {
		uint64_t max_factor = UINT64_MAX / current_base;
		
		if (exponent & 1) {
			if (result > max_factor) {
				return 0; // Overflow detected
			}
			result *= current_base;
		}

		exponent >>= 1;

		if (exponent > 0) {
			// Reuse the same division result for squaring check
			if (current_base > max_factor) {
				return 0; // Overflow detected
			}
			current_base *= current_base;
		}
	}

	return result;
}


@implementation NSNumber (Extension)

NSNumber* numberOperation(NSNumberMathOperation operation, NSNumber *first, NSNumber *second)
{
	// Define the precedence order of types
	static NSArray<NSString *> *typeOrder = nil;
	if (!typeOrder) {
#define kOperationWithChar				0
#define kOperationWithShort				1
#define kOperationWithInt				2
#define kOperationWithLong				3
#define kOperationWithLongLong			4
#define kOperationWithBool				5
#define kOperationWithUnsignedChar		6
#define kOperationWithUnsignedShort		7
#define kOperationWithUnsignedInt		8
#define kOperationWithUnsignedLong		9
#define kOperationWithUnsignedLongLong	10
#define kOperationWithFloat				11
#define kOperationWithDouble			12
		
		typeOrder = @[
			[NSString stringWithFormat:@"%s", @encode(char)],
			[NSString stringWithFormat:@"%s", @encode(short)],
			[NSString stringWithFormat:@"%s", @encode(int)],
			(strcmp(@encode(long), @encode(long long))) ? [NSString stringWithFormat:@"%s", @encode(long)] : @"l",
			[NSString stringWithFormat:@"%s", @encode(long long)],
			
			[NSString stringWithFormat:@"%s", @encode(BOOL)],
			
			[NSString stringWithFormat:@"%s", @encode(unsigned char)],
			[NSString stringWithFormat:@"%s", @encode(unsigned short)],
			[NSString stringWithFormat:@"%s", @encode(unsigned int)],
			(strcmp(@encode(unsigned long), @encode(unsigned long long))) ? [NSString stringWithFormat:@"%s", @encode(long)] : @"L",
			[NSString stringWithFormat:@"%s", @encode(unsigned long long)],
			
			[NSString stringWithFormat:@"%s", @encode(float)],
			[NSString stringWithFormat:@"%s", @encode(double)]
		];
	}

	// Access the characters of the objCType strings
	NSString *firstType = [NSString stringWithFormat:@"%s", [first objCType]];
	NSString *secondType = [NSString stringWithFormat:@"%s", [second objCType]];

	// Determine the index of the highest precedence type
	NSUInteger firstIndex = [typeOrder indexOfObject:firstType];
	NSUInteger secondIndex = [typeOrder indexOfObject:secondType];
	NSUInteger highestIndex = MAX(firstIndex, secondIndex);
	NSUInteger lowestIndex = MIN(firstIndex, secondIndex);
	
	if (highestIndex >= 5 && highestIndex <= 9) {
		highestIndex -= 5; //convert unsigned (5..9) to next largest signed, except for largest unsigned.
		NSUInteger newHighest = MAX(highestIndex, lowestIndex);
		NSUInteger newLowest = MIN(highestIndex, lowestIndex);
		highestIndex = newHighest;
		lowestIndex = newLowest;
		
		if (highestIndex >= 5 && highestIndex <= 9) {
			highestIndex -= 5; //convert unsigned (5..9) to next largest signed, except for largest unsigned.
			newHighest = MAX(highestIndex, lowestIndex);
			newLowest = MIN(highestIndex, lowestIndex);
			highestIndex = newHighest;
			lowestIndex = newLowest;
		}
	}
	
	NSNumber *resultNumber = nil;
	if (highestIndex <= 10) {
		// Convert both numbers to the highest precedence type
		BOOL _firstUnsigned = NO, _secondUnsigned = NO, hasUnsigned;
		
		UInt64 uFirstValue = 0;
		SInt64 firstValue = 0;
		if (firstIndex <= 4) {
			firstValue = [first longLongValue];
		} else {
			uFirstValue = [first unsignedLongLongValue];
			_firstUnsigned = YES;
		}
		
		UInt64 uSecondValue = 0;
		SInt64 secondValue = 0;
		if (secondIndex <= 4) {
			secondValue = [second longLongValue];
		} else {
			uSecondValue = [second unsignedLongLongValue];
			_secondUnsigned = YES;
		}
		hasUnsigned = _firstUnsigned | _secondUnsigned;
		
		// Perform the mathematical operation
		unsigned long long ullResult = 0;
		long long llResult = 0;
		switch (operation) {
			case NSNumberMathOperationAdd:
				if (_firstUnsigned && _secondUnsigned) {
					ullResult = uFirstValue + uSecondValue;
				} else if (_secondUnsigned) {
					ullResult = firstValue + uSecondValue;
				} else if (_firstUnsigned) {
					ullResult = uFirstValue + secondValue;
				} else {
					llResult = firstValue + secondValue;
				}
				break;
			case NSNumberMathOperationSubtract:
				if (_firstUnsigned && _secondUnsigned) {
					ullResult = uFirstValue - uSecondValue;
				} else if (_secondUnsigned) {
					ullResult = firstValue - uSecondValue;
				} else if (_firstUnsigned) {
					ullResult = uFirstValue - secondValue;
				} else {
					llResult = firstValue - secondValue;
				}
				break;
			case NSNumberMathOperationMultiply:
				if (_firstUnsigned && _secondUnsigned) {
					ullResult = uFirstValue * uSecondValue;
				} else if (_secondUnsigned) {
					ullResult = firstValue * uSecondValue;
				} else if (_firstUnsigned) {
					ullResult = uFirstValue * secondValue;
				} else {
					llResult = firstValue * secondValue;
				}
				break;
			case NSNumberMathOperationDivide:
				if ((_secondUnsigned && uSecondValue == 0) || (!_secondUnsigned && secondValue == 0)) {
					return @(NAN);
				}
				if (_firstUnsigned && _secondUnsigned) {
					ullResult = uFirstValue / uSecondValue;
				} else if (_secondUnsigned) {
					ullResult = firstValue / uSecondValue;
				} else if (_firstUnsigned) {
					ullResult = uFirstValue / secondValue;
				} else {
					llResult = firstValue / secondValue;
				}
				break;
			case NSNumberMathOperationModulus:
				if (_firstUnsigned && _secondUnsigned) {
					ullResult = uFirstValue % uSecondValue;
				} else if (_secondUnsigned) {
					ullResult = firstValue % uSecondValue;
				} else if (_firstUnsigned) {
					ullResult = uFirstValue % secondValue;
				} else {
					llResult = firstValue % secondValue;
				}
				break;
			case NSNumberMathOperationPower:
				if (_firstUnsigned && _secondUnsigned) {
					ullResult = pow_uint64(uFirstValue, uSecondValue);
				} else if (_secondUnsigned) {
					llResult = pow_int64(firstValue, uSecondValue);
					hasUnsigned = NO;
				} else if (_firstUnsigned) {
					ullResult = pow_uint64(uFirstValue, secondValue);
				} else {
					llResult = pow_int64(firstValue, secondValue);
				}
				break;
			case NSNumberMathOperationXor:
				if (_firstUnsigned && _secondUnsigned) {
					ullResult = uFirstValue ^ uSecondValue;
				} else if (_secondUnsigned) {
					ullResult = firstValue ^ uSecondValue;
				} else if (_firstUnsigned) {
					ullResult = uFirstValue ^ secondValue;
				} else {
					llResult = firstValue ^ secondValue;
				}
				break;
		}
		if (highestIndex == kOperationWithChar) {	// char
			resultNumber = [NSNumber numberWithChar:(char)(hasUnsigned ? ullResult : llResult)];
		} else if (highestIndex == kOperationWithShort) {
			resultNumber = [NSNumber numberWithShort:(short)(hasUnsigned ? ullResult : llResult)];
		} else if (highestIndex == kOperationWithInt) {
			resultNumber = [NSNumber numberWithInt:(int)(hasUnsigned ? ullResult : llResult)];
		} else if (highestIndex == kOperationWithLong) {
			resultNumber = [NSNumber numberWithLong:(long)(hasUnsigned ? ullResult : llResult)];
		} else if (highestIndex == kOperationWithLongLong) {
			resultNumber = [NSNumber numberWithLongLong:(long long)(hasUnsigned ? ullResult : llResult)];
		} else if (highestIndex == kOperationWithBool) {
			resultNumber = [NSNumber numberWithBool:(BOOL)(hasUnsigned ? ullResult : llResult)];
		} else if (highestIndex == kOperationWithUnsignedChar) {
			resultNumber = [NSNumber numberWithUnsignedChar:(unsigned char)(hasUnsigned ? ullResult : llResult)];
		} else if (highestIndex == kOperationWithUnsignedShort) {
			resultNumber = [NSNumber numberWithUnsignedShort:(unsigned short)(hasUnsigned ? ullResult : llResult)];
		} else if (highestIndex == kOperationWithUnsignedInt) {
			resultNumber = [NSNumber numberWithUnsignedInt:(unsigned int)(hasUnsigned ? ullResult : llResult)];
		} else if (highestIndex == kOperationWithUnsignedLong) {
			resultNumber = [NSNumber numberWithUnsignedLong:(unsigned long)(hasUnsigned ? ullResult : llResult)];
		} else if (highestIndex == kOperationWithUnsignedLongLong) {
			resultNumber = [NSNumber numberWithUnsignedLongLong:(unsigned long long)(hasUnsigned ? ullResult : llResult)];
		}
		
	} else {
		// Convert both numbers to the highest precedence type
		double firstValue = [first doubleValue];
		double secondValue = [second doubleValue];
		
		// Perform the mathematical operation
		double result = 0;
		switch (operation) {
			case NSNumberMathOperationAdd:
				result = firstValue + secondValue;
				break;
			case NSNumberMathOperationSubtract:
				result = firstValue - secondValue;
				break;
			case NSNumberMathOperationMultiply:
				result = firstValue * secondValue;
				break;
			case NSNumberMathOperationDivide:
				if (secondValue == 0.0) {
					return @(INFINITY);
				}
				result = firstValue / secondValue;
				break;
			case NSNumberMathOperationModulus:
				result = fmod(firstValue, secondValue);
				break;
			case NSNumberMathOperationPower:
				result = pow(firstValue, secondValue);
				break;
			case NSNumberMathOperationXor:
				result = (int)firstValue ^ (int)secondValue;
				break;
		}
		if (highestIndex == kOperationWithFloat) {
			resultNumber = [NSNumber numberWithFloat:(float)result];
		} else if (highestIndex == kOperationWithDouble) {
			resultNumber = [NSNumber numberWithDouble:result];
		}
	}

	return resultNumber;
}



- (NSNumber*)addNumber:(NSNumber *)second
{
	return numberOperation(NSNumberMathOperationAdd, self, second);
}

- (NSNumber*)subtractNumber:(NSNumber *)second
{
	return numberOperation(NSNumberMathOperationSubtract, self, second);
}

- (NSNumber*)multiplyNumber:(NSNumber *)second
{
	return numberOperation(NSNumberMathOperationMultiply, self, second);
}

- (NSNumber*)divideNumber:(NSNumber *)second
{
	return numberOperation(NSNumberMathOperationDivide, self, second);
}

- (NSNumber*)modulusNumber:(NSNumber *)second
{
	return numberOperation(NSNumberMathOperationModulus, self, second);
}

- (NSNumber*)powerNumber:(NSNumber *)second
{
	return numberOperation(NSNumberMathOperationPower, self, second);
}

- (NSNumber*)xorNumber:(NSNumber *)second
{
	return numberOperation(NSNumberMathOperationXor, self, second);
}





- (NSNumber*)addInt:(SInt64)second
{
	return numberOperation(NSNumberMathOperationAdd, self, [NSNumber numberWithLongLong:second]);
}

- (NSNumber*)addUInt:(UInt64)second
{
	return numberOperation(NSNumberMathOperationAdd, self, [NSNumber numberWithUnsignedLongLong:second]);
}

- (NSNumber*)addDouble:(double)second
{
	return numberOperation(NSNumberMathOperationAdd, self, [NSNumber numberWithDouble:second]);
}


- (NSNumber*)subtractInt:(SInt64)second
{
	return numberOperation(NSNumberMathOperationSubtract, self, [NSNumber numberWithLongLong:second]);
}
- (NSNumber*)subtractUInt:(UInt64)second
{
	return numberOperation(NSNumberMathOperationSubtract, self, [NSNumber numberWithUnsignedLongLong:second]);
}
- (NSNumber*)subtractDouble:(double)second
{
	return numberOperation(NSNumberMathOperationSubtract, self, [NSNumber numberWithDouble:second]);
}


- (NSNumber*)multiplyInt:(SInt64)second
{
	return numberOperation(NSNumberMathOperationMultiply, self, [NSNumber numberWithLongLong:second]);
}
- (NSNumber*)multiplyUInt:(UInt64)second
{
	return numberOperation(NSNumberMathOperationMultiply, self, [NSNumber numberWithUnsignedLongLong:second]);
}
- (NSNumber*)multiplyDouble:(double)second
{
	return numberOperation(NSNumberMathOperationMultiply, self, [NSNumber numberWithDouble:second]);
}


- (NSNumber*)divideInt:(SInt64)second
{
	return numberOperation(NSNumberMathOperationDivide, self, [NSNumber numberWithLongLong:second]);
}
- (NSNumber*)divideUInt:(UInt64)second
{
	return numberOperation(NSNumberMathOperationDivide, self, [NSNumber numberWithUnsignedLongLong:second]);
}
- (NSNumber*)divideDouble:(double)second
{
	return numberOperation(NSNumberMathOperationDivide, self, [NSNumber numberWithDouble:second]);
}


- (NSNumber*)modulusInt:(SInt64)second
{
	return numberOperation(NSNumberMathOperationModulus, self, [NSNumber numberWithLongLong:second]);
}
- (NSNumber*)modulusUInt:(UInt64)second
{
	return numberOperation(NSNumberMathOperationModulus, self, [NSNumber numberWithUnsignedLongLong:second]);
}
- (NSNumber*)modulusDouble:(double)second
{
	return numberOperation(NSNumberMathOperationModulus, self, [NSNumber numberWithDouble:second]);
}


- (NSNumber*)powerUInt:(UInt64)second
{
	return numberOperation(NSNumberMathOperationPower, self, [NSNumber numberWithUnsignedLongLong:second]);
}
- (NSNumber*)powerDouble:(double)second
{
	return numberOperation(NSNumberMathOperationPower, self, [NSNumber numberWithDouble:second]);
}


- (NSNumber*)xorInt:(SInt64)second
{
	return numberOperation(NSNumberMathOperationXor, self, [NSNumber numberWithLongLong:second]);
}
- (NSNumber*)xorUInt:(UInt64)second
{
	return numberOperation(NSNumberMathOperationXor, self, [NSNumber numberWithUnsignedLongLong:second]);
}

@end

File: Source/NSNumber+BExtension.swift
/*!
 @header		NSNumber+BExtension.swift
 @copyright		-© 2025 Delicense - @belisoful. All rights released.
 @date			2025-01-01
 @abstract
 @discussion	This is unfinished and unformed EXPERIMENTAL code for swift
*/

import Foundation

infix operator --
infix operator ++

public extension NSNumber {
	
	static func += (lhs: inout NSNumber, rhs: NSNumber) {
		lhs = lhs + rhs
	}
	
	static func -= (lhs: inout NSNumber, rhs: NSNumber) {
		lhs = lhs - rhs
	}
	
	static func *= (lhs: inout NSNumber, rhs: NSNumber) {
		lhs = lhs * rhs
	}
	
	static func /= (lhs: inout NSNumber, rhs: NSNumber) {
		lhs = lhs / rhs
	}
	
	static func %= (lhs: inout NSNumber, rhs: NSNumber) {
		lhs = lhs % rhs
	}
	
	static func ^= (lhs: inout NSNumber, rhs: NSNumber) {
		lhs = lhs ^ rhs
	}
}

@objc enum NSNumberMathOperation: Int {
	case add
	case subtract
	case multiply
	case divide
	case modulus
	case power
	case xor
}


private func numberOperation(operation: NSNumberMathOperation, first: NSNumber, second: NSNumber) -> NSNumber {
	// Define the precedence order of types
	let typeOrder: [UInt8] = [
		UInt8(ascii: "c"), UInt8(ascii: "s"), UInt8(ascii: "i"), UInt8(ascii: "l"), UInt8(ascii: "q"),
		
		UInt8(ascii: "B"),
		UInt8(ascii: "C"), UInt8(ascii: "S"), UInt8(ascii: "I"), UInt8(ascii: "L"), UInt8(ascii: "Q"),
		
		UInt8(ascii: "f"), UInt8(ascii: "d")
	]

	// Access the characters of the objCType strings
	let firstType = UInt8(first.objCType[0])
	let secondType = UInt8(second.objCType[0])

	// Determine the index of the highest precedence type
	let firstIndex = typeOrder.firstIndex(of: firstType) ?? Int.min
	let secondIndex = typeOrder.firstIndex(of: secondType) ?? Int.min
	let highestIndex = max(firstIndex, secondIndex)
	let highestType = typeOrder[highestIndex]

	var resultNumber: NSNumber = 0
	if (highestIndex <= 10) {
		var firstValue:UInt64
		if (firstIndex <= 4) {
			firstValue = UInt64(first.int64Value)
		} else {
			firstValue = first.uint64Value
		}
		
		var secondValue:UInt64
		if (secondIndex <= 4) {
			secondValue = UInt64(second.int64Value)
		} else {
			secondValue = second.uint64Value
		}
		var result: UInt64 = 0
		
		switch operation {
			case .add:
				result = firstValue + secondValue
			case .subtract:
				result = firstValue - secondValue
			case .multiply:
				result = firstValue * secondValue
			case .divide:
				result = (secondValue != 0) ? (firstValue / secondValue) : 0;
			case .modulus:
				result = firstValue % secondValue;
			case .power:
				result = UInt64(pow(Double(firstValue), Double(secondValue)))
			case .xor:
				result = firstValue ^ secondValue
		}
		if (highestType == UInt8(ascii: "c")) {
			resultNumber = NSNumber(value: Int8(result))
		} else if (highestType == UInt8(ascii: "C")) {
			resultNumber = NSNumber(value: UInt8(result))
		} else if (highestType == UInt8(ascii: "s")) {
			resultNumber = NSNumber(value: Int16(result))
		} else if (highestType == UInt8(ascii: "S")) {
			resultNumber = NSNumber(value: UInt16(result))
		} else if (highestType == UInt8(ascii: "i")) {
			resultNumber = NSNumber(value: Int32(result))
		} else if (highestType == UInt8(ascii: "I")) {
			resultNumber = NSNumber(value: UInt32(result))
		} else if (highestType == UInt8(ascii: "l")) {
			resultNumber = NSNumber(value: Int(result))
		} else if (highestType == UInt8(ascii: "L")) {
			resultNumber = NSNumber(value: UInt(result))
		} else if (highestType == UInt8(ascii: "q")) {
			resultNumber = NSNumber(value: Int64(result))
		} else if (highestType == UInt8(ascii: "Q")) {
			resultNumber = NSNumber(value: UInt64(result))
		}
	} else {
		let firstValue: Double = first.doubleValue
		let secondValue: Double = second.doubleValue
		
		var result: Double = 0
		
		switch operation {
		case .add:
			result = firstValue + secondValue
		case .subtract:
			result = firstValue - secondValue
		case .multiply:
			result = firstValue * secondValue
		case .divide:
			if secondValue != 0 {
				result = firstValue / secondValue
			} else {
				result = Double.nan
			}
		case .modulus:
			result = fmod(firstValue, secondValue)
		case .power:
			result = pow(firstValue, secondValue)
		case .xor:
			result = 0
		}
		if (highestType == UInt8(ascii: "f")) {
			resultNumber = NSNumber(value: Float(result))
		} else if (highestType == UInt8(ascii: "d")) {
			resultNumber = NSNumber(value: result)
		}
	}

	return resultNumber
}





///Add two numbers
public func + (first: NSNumber, second:NSNumber) -> NSNumber {
	return numberOperation(operation: .add, first: first, second: second)
}

///Subtract two numbers
public func - (first: NSNumber, second:NSNumber) -> NSNumber {
	return numberOperation(operation: .subtract, first: first, second: second)
}

///Multiply two numbers
public func * (first: NSNumber, second:NSNumber) -> NSNumber {
	return numberOperation(operation: .multiply, first: first, second: second)
}

///Divide two numbers
public func / (first: NSNumber, second:NSNumber) -> NSNumber {
	return numberOperation(operation: .divide, first: first, second: second)
}

///Divide two numbers
public func % (first: NSNumber, second:NSNumber) -> NSNumber {
	return numberOperation(operation: .modulus, first: first, second: second)
}

///Divide two numbers
public func pow (first: NSNumber, second:NSNumber) -> NSNumber {
	return numberOperation(operation: .power, first: first, second: second)
}
public func | (first: NSNumber, second:NSNumber) -> NSNumber {
	return true;
}
public func & (first: NSNumber, second:NSNumber) -> NSNumber {
	return true;
}
public func ^ (first: NSNumber, second:NSNumber) -> NSNumber {
	return true;
}

///Postfix increment two numbers
public prefix func ~(first: NSNumber) -> NSNumber {
	let type = UInt8(first.objCType[0])
	
	var number:UInt64
	switch type {
		case UInt8(ascii: "c"):
			number = UInt64(first.int8Value)
		case UInt8(ascii: "s"):
			number = UInt64(first.int16Value)
		case UInt8(ascii: "i"):
			number = UInt64(first.intValue)
		case UInt8(ascii: "l"):
			number = UInt64(first.int32Value)
		case UInt8(ascii: "q"):
			number = UInt64(first.int64Value)
		
		case UInt8(ascii: "B"):
			number = UInt64(first.uint8Value)
		case UInt8(ascii: "C"):
			number = UInt64(first.uint8Value)
		case UInt8(ascii: "S"):
			number = UInt64(first.uint8Value)
		case UInt8(ascii: "I"):
			number = UInt64(first.uint8Value)
		case UInt8(ascii: "L"):
			number = UInt64(first.uint8Value)
		case UInt8(ascii: "Q"):
			number = UInt64(first.uint8Value)
		
	default:
		number = 0
	}
	number = ~number;
	
	var resultNumber:NSNumber
	switch type {
		case UInt8(ascii: "c"):
			resultNumber = NSNumber(value: Int8(number))
		case UInt8(ascii: "s"):
			resultNumber = NSNumber(value: Int16(number))
		case UInt8(ascii: "i"):
			resultNumber = NSNumber(value: Int(number))
		case UInt8(ascii: "l"):
			resultNumber = NSNumber(value: Int32(number))
		case UInt8(ascii: "q"):
			resultNumber = NSNumber(value: Int64(number))
		
		case UInt8(ascii: "B"):
			resultNumber = NSNumber(value: UInt8(number))
		case UInt8(ascii: "C"):
			resultNumber = NSNumber(value: UInt8(number))
		case UInt8(ascii: "S"):
			resultNumber = NSNumber(value: UInt16(number))
		case UInt8(ascii: "I"):
			resultNumber = NSNumber(value: UInt(number))
		case UInt8(ascii: "L"):
			resultNumber = NSNumber(value: UInt32(number))
		case UInt8(ascii: "Q"):
			resultNumber = NSNumber(value: UInt64(number))
		
	default:
		resultNumber = 0
	}
	return resultNumber
}

public func < (first: NSNumber, second:NSNumber) -> ObjCBool {
	return ;
}
public func <= (first: NSNumber, second:NSNumber) -> ObjCBool {
	return true;
}
public func > (first: NSNumber, second:NSNumber) -> ObjCBool {
	return true;
}
public func >= (first: NSNumber, second:NSNumber) -> ObjCBool {
	return true;
}
public func == (first: NSNumber, second:NSNumber) -> ObjCBool {
	return true;
}
public func != (first: NSNumber, second:NSNumber) -> ObjCBool {
	return true;
}

///Postfix decrement two numbers
public prefix func -- (first: NSNumber) -> NSNumber {
	return numberOperation(operation: .subtract, first: first, second: 1)
}

///Postfix decrement two numbers
public postfix func -- (first: NSNumber) -> NSNumber {
	return numberOperation(operation: .subtract, first: first, second: 1)
}


public prefix func ++(first: NSNumber) -> NSNumber {
	return numberOperation(operation: .add, first: first, second: 1)
}

///Postfix increment two numbers
public postfix func ++(first: NSNumber) -> NSNumber {
	return numberOperation(operation: .add, first: first, second: 1)
}
